### PR TITLE
ARROW-5833: [C++] Factor out Status-enriching code

### PR DESCRIPTION
**Pull request by pitrou**

### CODE REVIEW

**Comment by pitrou** [comment]
@emkornfield

File: cpp/src/arrow/status.h [hunk 1]
```diff
@@ -303,12 +303,18 @@ class ARROW_EXPORT Status {
...
   /// updating with the existing detail.
-  Status WithDetail(std::shared_ptr<StatusDetail> new_detail) {
+  Status WithDetail(std::shared_ptr<StatusDetail> new_detail) const {
     return Status(code(), message(), std::move(new_detail));
   }
 
+  /// \brief Return a new Status with changed message, copying the
+  /// existing status code and detail.
+  Status WithMessage(std::string message) const {
+    return Status(code(), std::move(message), detail());
```
Comment (emkornfield) [inline cpp/src/arrow/status.h]: std::move here?

**Comment by emkornfield** [comment]
+1, LGTM can merge when CI completes.

File: cpp/src/arrow/status.h (hunk 1 above)
Comment (emkornfield) [reply cpp/src/arrow/status.h]: NM, already r-value

**Comment by kszucs** [comment]
Appveyor R error is  unrelated.

**Review by kszucs — APPROVED** [review]

**Pull request closed without merging**